### DISCUSSION TITLE
I created an extension that adds a task, but vscode does not recognise it

### DISCUSSION THREAD

**Discussion by BlackRam-oss** [microsoft/vscode-discussions#718] [Extension Development QnA] [answered]
I am trying to create an extension that adds a task to run games written in renpy

I followed this guide: https://code.visualstudio.com/api/extension-guides/task-provider

But I have been plotted on a problem for a long time: it gives me the following error: there are no registered activities of type 'renpy'.     
![image](https://github.com/microsoft/vscode-discussions/assets/67595890/005230c4-c811-431e-a39c-941ecf1cca54)

what did i do wrong? i don't understand. 

the implementation is as follows:

(https://github.com/DRincs-Productions/renpy-vscode/tree/3-run-the-command-from-the-system-terminal-debugger)

```ts
export function activate(context: vscode.ExtensionContext) {
	// ...
	console.log("Ren'Py VSCode extension activated!");

	// register task provider
	context.subscriptions.push(registerTaskProvider());

        // ...
}
```

```ts
import * as vscode from 'vscode';
import { Disposable } from '../utils/dispose';

// https://code.visualstudio.com/api/extension-guides/task-provider

class RenpyTaskDefinition implements vscode.TaskDefinition {
	constructor(public task: string) {
		this.task = task;
	}
	[name: string]: any;
	type: string = 'renpy'
}

/**
 * Provides tasks for building `tsconfig.json` files in a project.
 */
class RenpyTaskProvider extends Disposable implements vscode.TaskProvider {
	public constructor() {
		super();
		this._register(vscode.workspace.onDidChangeConfiguration(this.onConfigurationChanged, this));
		this.onConfigurationChanged();
	}

	public async provideTasks(token: vscode.CancellationToken): Promise<vscode.Task[]> {
		return Promise.resolve([
			new vscode.Task(
				new RenpyTaskDefinition('run'),
				vscode.TaskScope.Workspace,
				'Run',
				'renpy',
			),
			new vscode.Task(
				new RenpyTaskDefinition('compile'),
				vscode.TaskScope.Workspace,
				'Re-Compile',
				'renpy',
			),
			new vscode.Task(
				new RenpyTaskDefinition('rmpersistent'),
				vscode.TaskScope.Workspace,
				'Delete Persistent',
				'renpy',
			),
			new vscode.Task(
				new RenpyTaskDefinition('lint'),
				vscode.TaskScope.Workspace,
				'Lint',
				'renpy',
			),
			new vscode.Task(
				new RenpyTaskDefinition('distribute'),
				vscode.TaskScope.Workspace,
				'Distribute',
				'renpy',
			),
		]);
	}

	public async resolveTask(_task: vscode.Task): Promise<vscode.Task | undefined> {
		if (!process.env.hasOwnProperty('RenPy')) {
			vscode.window.showErrorMessage("Ren'Py SDK not found. Please set the environment variable RenPy to the path of your Ren'Py SDK.");
			return undefined;
		}

		let renpy = process.env['RenPy'];
		if (renpy === undefined) {
			vscode.window.showErrorMessage("Ren'Py SDK not found. Please set the environment variable RenPy to the path of your Ren'Py SDK.");
			return undefined;
		}

		let executable = undefined;
		if (process.platform === 'win32') {
			executable = renpy + '\\renpy.exe';
		} else if (process.platform === 'linux') {
			executable = renpy + '/renpy.sh';
		} else if (process.platform === 'darwin') {
			executable = renpy + '/renpy.sh';
		} else {
			vscode.window.showErrorMessage("Platform not supported.")
			return undefined;
		}

		const task = _task.definition.task;
		// A Renpy task consists of a task and an optional file as specified in RenpyTaskDefinition
		// Make sure that this looks like a Renpy task by checking that there is a task.
		if (task) {
			// resolveTask requires that the same definition object be used.
			const definition: RenpyTaskDefinition = <any>_task.definition;
			return new vscode.Task(
				definition,
				_task.scope ?? vscode.TaskScope.Workspace,
				definition.task,
				definition.type,
				new vscode.ProcessExecution(executable, [definition.task]), // new ShellExecution
			);
		}
		return undefined;
	}

	private onConfigurationChanged(): void {
		// const type = vscode.workspace.getConfiguration('typescript.tsc').get<AutoDetect>('autoDetect');
		// this.autoDetect = typeof type === 'undefined' ? AutoDetect.on : type;
	}
}

export function registerTaskProvider() {
	return vscode.tasks.registerTaskProvider('renpy', new RenpyTaskProvider());
}

```

json: https://github.com/DRincs-Productions/renpy-vscode/blob/3-run-the-command-from-the-system-terminal-debugger/package.json

```json
 "activationEvents": [
    "workspaceContains:**/*.rpy",
    "onTaskType:renpy"
  ],
```

```json
      "taskDefinitions": [
        {
          "type": "renpy",
          "required": [
            "task"
          ],
          "properties": {
            "task": {
              "type": "string",
              "description": "The Renpy task to customize"
            }
          }
        }
      ]
```

**Comment by r3m0t** [comment]
Did your extension get activated? Add breakpoints and check the Output window (Extension Host logs)

**Reply by BlackRam-oss** [reply]
Yes the extension is active passes in the all code, but not in the following functions:      
- `provideTasks(token: vscode.CancellationToken)`      
- `resolveTask(_task: vscode.Task)`       


and there are no error logs

**Reply by r3m0t** [reply]
I don't know then :(

**Reply by isidorn** [reply]
Maybe @meganrogge can help here.

**Reply by meganrogge** [reply]
I'd suggest checking the task output for errors / dev tools console also if you haven't already.

**Reply by meganrogge** [reply]
and you run `Run tasks` and it's not there?

**Comment by BlackRam-oss** [comment]
Hi @meganrogge  ,              
Unfortunately, it is not possible to run the task because vscode crashes me first.          

 these are all the outputs I can think of:           
![image](https://github.com/microsoft/vscode-discussions/assets/67595890/1b4ccbeb-e676-4c27-a328-50dd092f5c85)                 
![image](https://github.com/microsoft/vscode-discussions/assets/67595890/6627530f-9bce-41ff-83fc-71a61e2ccf2b)                  
![image](https://github.com/microsoft/vscode-discussions/assets/67595890/0c9443c8-2631-4775-a64c-1a2b17e60045)         
![image](https://github.com/microsoft/vscode-discussions/assets/67595890/ce2e8855-e516-4bbf-b183-7bb656086809)               


i noticed that it tells me that the `process` parameter is invalid, but i don't declare it. i don't understand.            

please help me

**Comment by meganrogge** [comment]
>Unfortunately, it is not possible to run the task because vscode crashes me first.

pls create an issue to address this

**Reply by BlackRam-oss** [reply]
Sorry for responding so much later.

By that sentence I didn't mean that vscode closes or gives some error.

I meant that simply that task is not available

I open an issues?

**Comment by BlackRam-oss** [comment]
@meganrogge 
I think the only way to figure out the error is to test the extension. I can do the tests you ask, but I'm afraid it's something stupid that I don't see
the branch is as follows: https://github.com/DRincs-Productions/renpy-vscode/tree/3-run-the-command-from-the-system-terminal-debugger

and the extension is triggered if the open host has a file **/*.rpy and "onTaskType:renpy"

**Comment by BlackRam-oss** [comment]
please help me

**Comment by meganrogge** [comment, marked as answer]
this issue is tracked here and i'll try to get to it in september https://github.com/microsoft/vscode/issues/191745
